lib/stinger_core/inc/stinger_batch_insert.h

```diff
@@ -0,0 +1,571 @@
+template<typename adapter, typename iterator>
+void stinger_batch_insert_edges(stinger_t * G, iterator begin, iterator end);
+template<typename adapter, typename iterator>
+void stinger_batch_incr_edge_pairs(stinger_t * G, iterator begin, iterator end);
+template<typename adapter, typename iterator>
+void stinger_batch_insert_edge_pairs(stinger_t * G, iterator begin, iterator end);
+
+// *** Implementation ***
+namespace gt { namespace stinger {
+
+/*
+ * Rather than add these template arguments to every function in the file, just wrap everything in a template class
+ *
+ * adapter - provides static methods for accessing fields of an update
+ * iterator - iterator for a collection of updates
+ */
+template<typename adapter, typename iterator>
+class BatchInserter
+{
+protected:
+    // Everything in this class is protected and static, these friend functions are the only public interface
+    BatchInserter() {}
+    friend void stinger_batch_incr_edges<adapter, iterator>(stinger_t *G, iterator begin, iterator end);
+    friend void stinger_batch_insert_edges<adapter, iterator>(stinger_t * G, iterator begin, iterator end);
+    friend void stinger_batch_incr_edge_pairs<adapter, iterator>(stinger_t * G, iterator begin, iterator end);
+    friend void stinger_batch_insert_edge_pairs<adapter, iterator>(stinger_t * G, iterator begin, iterator end);
+
+    // what 'iterator' points to
+    typedef typename std::iterator_traits<iterator>::value_type update;
+
+    // Result codes
+    // Before returning to caller, we subtract 10 to make return codes match up with functions like stinger_incr_edge
+    enum result_codes {
+        PENDING = 0,
+        EDGE_ADDED = 11,    // 1
+        EDGE_UPDATED = 10,  // 0
+        EDGE_NOT_ADDED = 9  // -1
+    };
+
+    // Set of functions to sort/filter a collection of updates by source vertex
+    struct source_funcs
+    {
+        static int64_t
+        get(const update &x){
+            return adapter::get_source(x);
+        }
+        static void
+        set(update &x, int64_t neighbor){
+            adapter::set_source(x, neighbor);
+        }
+        static bool
+        equals(const update &a, const update &b){
+            return adapter::get_source(a) == adapter::get_source(b);
+        }
+        static bool
+        compare(const update &a, const update &b){
+            return adapter::get_source(a) < adapter::get_source(b);
+        }
+        static bool
+        sort(const update &a, const update &b){
+            return (adapter::get_type(a) != adapter::get_type(b)) ? adapter::get_type(a) < adapter::get_type(b) :
+                   (adapter::get_source(a) != adapter::get_source(b)) ? adapter::get_source(a) < adapter::get_source(b) :
+                   (adapter::get_dest(a) != adapter::get_dest(b)) ? adapter::get_dest(a) < adapter::get_dest(b) :
+                   (adapter::get_time(a) != adapter::get_time(b)) ? adapter::get_time(a) < adapter::get_time(b) :
```

Comment:
My preference would be to define a function that looks like the following. No need to nest. I could be alone in this opinion.

```
int sort_source_dest(const update &a, const update &b){
	if (adapter::get_type(a) != adapter::get_type(b))
		return adapter::get_type(a) < adapter::get_type(b);
	
	if (adapter::get_source(a) != adapter::get_source(b)) 
		return adapter::get_source(a) < adapter::get_source(b);

	if (adapter::get_dest(a) != adapter::get_dest(b))
		return adapter::get_dest(a) < adapter::get_dest(b);

	if (adapter::get_time(a) != adapter::get_time(b))
		return adapter::get_time(a) < adapter::get_time(b);
	return false;
}
```